src/briefcase/platforms/android/apk.py

```diff
@@ -1 +1,235 @@
+    description = "Create and populate an Android APK."
+
+
+class ApkUpdateCommand(ApkMixin, UpdateCommand):
+    description = "Update an existing Android APK."
+
+
+class ApkBuildCommand(ApkMixin, BuildCommand):
+    description = "Build an Android APK."
+
+    @property
+    def android_sdk_url(self):
+        # TODO: Add test validating that, if this is mocked out for a sentinel
+        # ZIP file, we surely unpack it into android_sdk_path.
+        """The Android SDK URL appropriate to this operating system."""
+        # The URLs described by the pattern below have existed since
+        # approximately 2017, and the code they download has a built-in
+        # updater. I hope they will work for many years.
+        return "https://dl.google.com/android/repository/" + (
+            "sdk-tools-{os}-4333796.zip".format(os=self.host_os.lower()))
+
+    def build_app(self, app: BaseConfig, **kwargs):
+        """
+        Build an application.
+
+        :param app: The application to build
+        """
+        print("[{app.app_name}] Building Android APK...".format(app=app))
+
+        try:
+            self.subprocess.run(
+                ["./gradlew", "assembleDebug"],
+                env=dict(list(self.os.environ.items()) + [
+                    ('ANDROID_SDK_ROOT', str(self.android_sdk_path))]),
+                check=True,
+                cwd=str(self.bundle_path(app)),
+            )
+        except subprocess.CalledProcessError:
+            # TODO: Capture and print gradle log, in case of error.
+            raise BriefcaseCommandError(
+                "Error while building app {app.app_name}.".format(app=app)
+            )
+
+
+NO_OR_WRONG_DEVICE_MESSAGE = """\
+You can get a list of valid devices by running this command and looking in the
+first column of output.
+
+$ {adb} devices -l
+
+If you do not see any devices, you can create one by running these commands:
+
+$ {tools_bin}/sdkmanager "platforms;android-28" \
+    "system-images;android-28;default;x86" "emulator" "platform-tools"
+
+$ {tools_bin}/avdmanager --verbose create avd --name robotfriend \
+    --abi x86 --package 'system-images;android-28;default;x86' --device pixel
+
+$ {emulator} -avd robotfriend &
+
+Then use adb find out the device name by running this command and looking
+in the first column of output.
+
+$ {adb} devices -l
+"""
+
+
+class ApkRunCommand(ApkMixin, RunCommand):
+    description = "Run an Android APK."
+
+    def verify_tools(self):
+        super().verify_tools()
+        if not (self.android_sdk_path / "emulator").exists():
+            print("Ensuring we have the Android emulator and system image...")
+            # TODO: Error handling.
+            self.subprocess.run([
+                self.android_sdk_path / "tools" / "bin" / "sdkmanager",
+                "platforms;android-28",
+                "system-images;android-28;default;x86",
+                "emulator",
+                "platform-tools",
+            ])
```

Comment:
Will this hit the network on every execution? Is there any way to shortcut this by verifying the existence of files that should exist (e.g., checking for the actual binaries we need?)

Reply:
Good question. I'll investigate.

Reply:
See https://github.com/beeware/briefcase/pull/321/files/3a69bb84d382ec775678c7305808c0f20751540d#diff-fe2ea694d155bb0bf9a20276355d898bR155 (line 155) -- we only do this if `emulator` doesn't exist. For that reason, I think it's peaceful as-is.